chapters/en/slides/module6/module6_02.Rmd

```diff
@@ -0,0 +1,352 @@
+
+---
+
+## Coloring individual regions differently
+
+```{python}
+(alt.Chart(state_map).mark_geoshape().encode(
+    color='id:Q')
+.project(type='albersUsa'))
+```
+
+Notes:
+So far we have seen how to visualize region in a uniform color,
+but often we want to use color to represent amounts within regions
+so that we can contrast them with each other.
+How can we do this in Altair?
+
+As you can see in this slide,
+we can use `encode` together with `mark_geojson`,
+the same way we have used it with the other marks.
+This type of visualization can be thought of as a geographical heatmap
+and is often referred to as a "choropleth map".
+
+The `'id'` field here comes from the sample topojson file.
+If we opened up and looked at this file,
+we would find that it does contains a field called `'id'`,
+which is a numerical identifier for each state 
+in alphabetical order.
+
+Because we are not using a pandas dataframe,
+we mush specify the type of data in the column.
+Here we are specifying it a quantitative for demonstration purposes
+because the legend would become very long
+if we specified it as ordinal or nominal.
+
+---
+
+## Visualizing data from another dataframe on a map
+
+```{python}
+import pandas as pd
+
+state_pop = pd.read_csv('data/us_population_and_coordinates.csv')
+state_pop[:5]
+```
+
+Notes:
+So far we have learned how to visualize maps,
+but we haven't really been able to ask and answer any data question.
+
+Although any arbitrary information could be put in a geojson file,
+they often only contain spatial coordinates
+and we need to link any variables of interest from another data set
+together with the JSON file.
+
+For example,
+we might have this table with the population for each state.
+So how do we link these two tables together?
+
+---
+
+## Linking datasets together via lookups
+
+```{python}
+(alt.Chart(state_map).mark_geoshape().transform_lookup(
+    lookup='id',
+    from_=alt.LookupData(state_pop, 'id', ['population']))
+.encode(color='population:Q')
+.project(type='albersUsa'))
+```
+
+Notes:
+We could perform the linkage manually,
+by adding the relevant fields to the json file,
+but this would be a tedious process.
+
+There are some libraries,
+such as geopandas
+which can ameliorate this situation,
+but we could also use Altair directly to save time.
+
+To link two datasets together,
+Altair uses the `transform_lookup` method
+to *look up* the data column we want to use from the dataframe
+via a shared key column
+and add it to the plotted data.
+This process is similar to using `merge` in pandas.
+
+You could see in the previous slide that there exists an ID column
+and this contain the state IDs in alphabetical order,
+just like the geojson so we can use that for the lookup.
+
+The columns we want to fetch from the dataframe are passed in a list,
+and here we have specified only the population,
+which we use to color the states accordingly to how many people live in each state.
+
```

Comment:
```suggestion
We could perform the linkage manually,
by adding the relevant fields to the json file,
but this would be a tedious and error-prone process.

There are some libraries,
such as geopandas
which can ameliorate this situation,
but we could also use Altair directly to save time.

To link two datasets together,
Altair uses the `transform_lookup` method
to *look up* the data column we want to use from the dataframe
via a shared key column.
And then adds it to the plotted data.
This process is similar to using `merge` in pandas.

You could see in the previous slide that there exists an ID column
and this contains the state IDs in alphabetical order,
just like the topoJSON file - and so we can use that for the lookup.

The columns we want to fetch from the data frame are passed in a list,
and here we have specified only the population, We will use that to 
colour the states accordingly to how many people live in each state.

```

Reply:
I wonder why gits worddiff keeps failing on these section, it's hard to see what has changes when it displays them like this